File: examples/java/src/main/java/PaymentWorkflow.java
import java.util.concurrent.TimeUnit;
import java.util.*;

import com.ripple.payments.direct.*;
import com.ripple.payments.direct.api.*;
import com.ripple.payments.direct.auth.*;
import com.ripple.payments.direct.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PaymentWorkflow {

    private static final String clientId = "";
    private static final String clientSecret = "";
    private static final String baseUrl = "";
    private static final String tenantId = "urn:ripplexcurrent-<env>:<tenantName>";
    private static String quoteId;
    private static String identityId;
    private static AuthenticationApi authApi;
    private static QuoteApi quoteApi;
    private static IdentitiesV2Api identityApi;
    private static PaymentsApi paymentsApi;
    private static LedgerPublicApi ledgerApi;
    private static Logger logger;


    public static void setup() {
        ApiClient apiClient = new ApiClient();
        apiClient.setBasePath(baseUrl);

        authApi = new AuthenticationApi(apiClient);
        logger = LoggerFactory.getLogger(PaymentWorkflow.class);

    }

    // Get JWT token
    public static void authenticate() throws ApiException {

        AuthenticationRequestDTO authRequest = new AuthenticationRequestDTO()
                .clientId(clientId)
                .clientSecret(clientSecret)
                .audience(tenantId)
                .grantType(AuthenticationRequestDTO.GrantTypeEnum.CLIENT_CREDENTIALS);


        AuthenticationResponseDTO response = authApi.authenticate(authRequest, null);

        String accessToken = response.getAccessToken();
        logger.info("Access Token received.");

        // Update clients with auth
        ApiClient authClient = new ApiClient();
        authClient.setBasePath(baseUrl);
        authClient.setBearerToken(accessToken);

        quoteApi = new QuoteApi(authClient);
        identityApi = new IdentitiesV2Api(authClient);
        paymentsApi = new PaymentsApi(authClient);
        ledgerApi = new LedgerPublicApi(authClient);

    }

    // Get a quote
    public static void getQuote() throws ApiException {
        QuoteCollectionRequestDTO quoteRequest = new QuoteCollectionRequestDTO()
                .sourceCountry("US")
                .sourceCurrency("USD")
                .destinationCurrency("MXN")
                .destinationCountry("MX")
                .quoteAmount(Double.parseDouble("20"))
                .quoteAmountType(QuoteAmountTypeDTO.valueOf("SOURCE_AMOUNT"))
                .payinCategory("T_PLUS_ONE")
                .payoutCategory("BANK");


        QuoteCollectionDTO response = quoteApi.createQuoteCollection(quoteRequest);
        quoteId = String.valueOf(response.getQuotes().get(0).getQuoteId());
        logger.info("Quote ID: {}", quoteId);

    }

    // Create a beneficiary
    public static void createBeneficiary() throws ApiException {
        Map<String, Object> beneficiary = new HashMap<>();
        beneficiary = buildPiiData();

        CreateIdentityRequestV2DTO identityRequestV2 = new CreateIdentityRequestV2DTO();
        identityRequestV2.setNickName("Lindgren - Breitenberg");
        identityRequestV2.setIdentityType(IdentityTypeV2DTO.BENEFICIARY);
        identityRequestV2.setUseCaseType(UseCaseTypeDTO.BUSINESS);
        identityRequestV2.setPiiData(beneficiary);

        CreateIdentityResponseDTO response = identityApi.createIdentityV2(identityRequestV2);
        identityId = String.valueOf(response.getIdentityId());
        logger.info("Beneficiary Identity ID: {}", identityId);

        IdentityResponseV2DTO getIdentityResponse = identityApi.getIdentityByIdV2(identityId, null);
        logger.info("Beneficiary Identity ID using getIdentityById: {}", getIdentityResponse.getIdentityId());

    }

    // Create a payment
    public static void createPayment() throws ApiException {
        PaymentRequestDTO paymentRequest = new PaymentRequestDTO()
                .beneficiaryIdentityId(UUID.fromString(identityId))
                .purposeCode("OTHR")
                .internalId("test")
                .paymentLabels(Collections.singletonList("test-payment"))
                .quoteId(UUID.fromString(quoteId));

        PaymentDTO response = paymentsApi.createPayment(paymentRequest);
        logger.info("Payment ID: {}", response.getPaymentId());
    }

    // Poll for terminal payment state
    public static void waitForCompletedState() throws Exception {
        int maxAttempts = 25;
        int delaySeconds = 10;

        try {
            for (int i = 0; i < maxAttempts; i++) {
                StateTransitionsResponseDTO transitions = paymentsApi.getPaymentStateTransitionsById(UUID.fromString(quoteId));
                List<StateTransitionDTO> states = transitions.getStateTransitions();

                if (states.stream().anyMatch(t -> "COMPLETED".equals(String.valueOf(t.getUpdatedTo())))) {
                    logger.info("Payment completed successfully: {}", states);
                    return;
                }

                if (states.stream().anyMatch(t -> "FAILED".equals(String.valueOf(t.getUpdatedTo())))) {
                    PaymentWithDetailsDTO payment = paymentsApi.getPaymentById(UUID.fromString(quoteId));
                    List<PaymentErrorDTO> errors = payment.getErrors();
                    if (errors != null && !errors.isEmpty()) {
                        PaymentErrorDTO err = errors.get(0);
                        String errorMessage = "code:" + err.getCode() + "; type" + err.getType() + "; title:" + err.getTitle() + "; description:" + err.getDescription();
                        throw new Exception("Payment reached FAILED state. First error: " + errorMessage);
                    }
                    throw new Exception("Payment reached FAILED state with unknown error.");
                }

                logger.info("Attempt {}: Payment not yet completed or failed.", i + 1);
                TimeUnit.SECONDS.sleep(delaySeconds);
            }
        } catch (Exception e) {
            logger.info("Error polling payment status: {}", e);
            throw new Exception();
        }

        throw new Exception("Payment did not reach a terminal state within the allowed number of attempts.");
    }

    // Delete benecifiary
    public static void deleteBeneficiary() throws ApiException {

        identityApi.deactivateIdentityV2(identityId);
        logger.info("Beneficiary is deactivated/deleted.");

    }

    // If pre-funded, then use this. Otherwise this api is not needed.
    public static void getBalance() throws ApiException {

        ledgerApi.getBalances("USD");

    }

    public static void main(String[] args) throws Exception {
        setup();
        authenticate();
        getQuote();
        createBeneficiary();
        createPayment();
        waitForCompletedState();
        deleteBeneficiary();
        getBalance();
    }


    public static Map<String, Object> buildPiiData() {
        Map<String, Object> piiData = new HashMap<>();

        // --- Cdtr ---
        Map<String, Object> cdtr = new HashMap<>();
        cdtr.put("Nm", "Botsford and Sons");

        Map<String, Object> ctctDtls = new HashMap<>();
        ctctDtls.put("MobNb", "16-834-748-1777");
        ctctDtls.put("EmailAdr", "Raoul94@yahoo.com");
        cdtr.put("CtctDtls", ctctDtls);

        Map<String, Object> pstlAdr = new HashMap<>();
        pstlAdr.put("AdrLine", List.of("3103 Orn Garden"));
        pstlAdr.put("PstCd", "75357");
        pstlAdr.put("TwnNm", "West Marjolaine");
        pstlAdr.put("CtrySubDvsn", "Durango");
        pstlAdr.put("Ctry", "MX");
        cdtr.put("PstlAdr", pstlAdr);

        Map<String, Object> id = new HashMap<>();
        Map<String, Object> orgId = new HashMap<>();
        Map<String, Object> othr = new HashMap<>();
        Map<String, Object> schmeNm = new HashMap<>();
        Map<String, Object> cntrIncor = new HashMap<>();
        cntrIncor.put("CtryofIncorprtn", "MX");
        schmeNm.put("Cd", "TXID");

        othr.put("Id", "751-539-9808");
        othr.put("SchmeNm", schmeNm);
        orgId.put("Othr", othr);
        orgId.put("RelShipToDbtr", "SUBS");
        orgId.put("dtAndPlcofIncorp", cntrIncor);
        id.put("OrgId", orgId);
        cdtr.put("Id", id);

        piiData.put("Cdtr", cdtr);

        // --- CdtrAcct ---
        Map<String, Object> cdtrAcct = new HashMap<>();
        cdtrAcct.put("Ccy", "MXN");

        Map<String, Object> acctOthr = new HashMap<>();
        Map<String, Object> acctSchmeNm = new HashMap<>();
        acctSchmeNm.put("Cd", "CLABE");
        acctOthr.put("SchmeNm", acctSchmeNm);
        acctOthr.put("Id", "002010077777777771");

        Map<String, Object> acctId = new HashMap<>();
        acctId.put("Othr", acctOthr);
        cdtrAcct.put("Id", acctId);

        piiData.put("CdtrAcct", cdtrAcct);

        // --- CdtrAgt ---
        Map<String, Object> cdtrAgt = new HashMap<>();

        Map<String, Object> brnchId = new HashMap<>();
        Map<String, Object> brnchPstlAdr = new HashMap<>();
        brnchPstlAdr.put("Ctry", "MX");
        brnchId.put("PstlAdr", brnchPstlAdr);

        Map<String, Object> finInstnId = new HashMap<>();
        finInstnId.put("Nm", "enhance");

        cdtrAgt.put("BrnchId", brnchId);
        cdtrAgt.put("FinInstnId", finInstnId);

        piiData.put("CdtrAgt", cdtrAgt);

        return piiData;
    }
}

File: examples/python/payment_workflow.py
import time
import logging

from ripple_payments_direct.configuration import Configuration
from ripple_payments_direct.api_client import ApiClient
from ripple_payments_direct.api.authentication_api import AuthenticationApi
from ripple_payments_direct.api.quote_api import QuoteApi
from ripple_payments_direct.api.identities_v2_api import IdentitiesV2Api
from ripple_payments_direct.api.payments_api import PaymentsApi
from ripple_payments_direct.api.ledger_public_api import LedgerPublicApi
from ripple_payments_direct.models import *

# Setup logger
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("PaymentWorkflow")

# Configuration
client_id = ""
client_secret = ""
base_url = ""
tenant_id = ""

# Globals
quote_id = None
identity_id = None

def setup_api_client():
    config = Configuration(host=base_url)
    return ApiClient(configuration=config)

# Get JWT token
def authenticate():
    global quote_api, identity_api, payments_api, ledger_api

    api_client = setup_api_client()
    auth_api = AuthenticationApi(api_client)

    auth_request = AuthenticationRequest(
        client_id=client_id,
        client_secret=client_secret,
        audience=tenant_id,
        grant_type="client_credentials"
    )

    response = auth_api.authenticate(auth_request)
    access_token = response.access_token
    logger.info("Authenticated successfully.")

    # Set auth token for other APIs
    api_client.configuration.access_token = access_token

    quote_api = QuoteApi(api_client)
    identity_api = IdentitiesV2Api(api_client)
    payments_api = PaymentsApi(api_client)
    ledger_api = LedgerPublicApi(api_client)

# Get a quote
def get_quote():
    global quote_id

    request = QuoteCollectionRequest(
        source_country="US",
        source_currency="USD",
        destination_currency="MXN",
        destination_country="MX",
        quote_amount=20.0,
        quote_amount_type=QuoteAmountType("SOURCE_AMOUNT"),
        payin_category="T_PLUS_ONE",
        payout_category="BANK"
    )

    response = quote_api.create_quote_collection(request)
    quote_id = str(response.quotes[0].quote_id)
    logger.info(f"Quote ID: {quote_id}")


def build_pii_data():
    return {
        "Cdtr": {
            "Nm": "Botsford and Sons",
            "CtctDtls": {
                "MobNb": "16-834-748-1777",
                "EmailAdr": "Raoul94@yahoo.com"
            },
            "PstlAdr": {
                "AdrLine": ["3103 Orn Garden"],
                "PstCd": "75357",
                "TwnNm": "West Marjolaine",
                "CtrySubDvsn": "Durango",
                "Ctry": "MX"
            },
            "Id": {
                "OrgId": {
                    "RelShipToDbtr": "SUBS",
                    "dtAndPlcofIncorp": {"CtryofIncorprtn": "MX"},
                    "Othr": {
                        "Id": "751-539-9808",
                        "SchmeNm": {"Cd": "TXID"}
                    }
                }
            }
        },
        "CdtrAcct": {
            "Ccy": "MXN",
            "Id": {
                "Othr": {
                    "Id": "002010077777777771",
                    "SchmeNm": {"Cd": "CLABE"}
                }
            }
        },
        "CdtrAgt": {
            "FinInstnId": {"Nm": "enhance"},
            "BrnchId": {
                "PstlAdr": {"Ctry": "MX"}
            }
        }
    }

# Create a beneficiary
def create_beneficiary():
    global identity_id

    request = CreateIdentityRequestV2(
        nick_name="Lindgren - Breitenberg",
        identity_type=IdentityTypeV2("BENEFICIARY"),
        use_case_type=UseCaseType("BUSINESS"),
        pii_data=build_pii_data()
    )

    response = identity_api.create_identity_v2(request)
    identity_id = str(response.identity_id)
    logger.info(f"Beneficiary Identity ID: {identity_id}")

# Create a payment
def create_payment():
    request = PaymentRequest(
        beneficiary_identity_id=identity_id,
        purpose_code="OTHR",
        internal_id="test",
        payment_labels=["test-payment"],
        quote_id=quote_id
    )

    response = payments_api.create_payment(request)
    payment_id = str(response.payment_id)
    logger.info(f"Payment ID: {payment_id}")

# Poll for terminal payment state
def poll_payment_state(max_attempts=25, delay_seconds=10):
    for i in range(max_attempts):
        transitions = payments_api.get_payment_state_transitions_by_id(quote_id)
        states = transitions.state_transitions

        for state in states:
            updated_to = state.updated_to
            if updated_to == PaymentState.COMPLETED:
                logger.info("Payment completed successfully.")
                return
            if updated_to == PaymentState.FAILED:
                logger.error("Payment failed.")
                raise Exception("Payment failed.")

        logger.info(f"Attempt {i + 1}: Payment not yet completed or failed.")
        time.sleep(delay_seconds)

    raise TimeoutError("Payment did not reach a terminal state within the allowed number of attempts.")

# Delete beneficiary
def delete_beneficiary():
    identity_api.deactivate_identity_v2(identity_id)
    logger.info("Beneficiary deleted.")

# If pre-funded, then use this. Otherwise, this api is not needed.
def get_balance():
    ledger_api.get_balances("USD")


if __name__ == "__main__":
    authenticate()
    get_quote()
    create_beneficiary()
    create_payment()
    poll_payment_state()
    delete_beneficiary()
    get_balance()
